### PR TITLE
fix disjoints in query snippet

### PR DESCRIPTION
**Pull request by hkernbach**
add disjoint smart graph query snippet support

### CODE REVIEW

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/9679/

File: arangod/Aql/QuerySnippet.cpp [hunk 1]
```diff
@@ -225,8 +224,12 @@ void QuerySnippet::serializeIntoBuilder(
           TRI_ASSERT(colAcc != nullptr);
           if (colAcc->collection() == distCollection) {
             // Found one, use all shards of it
-            for (auto const& s : exp.second) {
-              distIds.emplace_back(s);
+            // unordered_map of collections to sets of ids
+            for (auto const& collectionMap : exp.second) {
+              // set of shard ids
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:229]: Please assert that collectionMap only has 1 entry in this case.
And that this entry is equal to colAcc->collection()

File: arangod/Aql/QuerySnippet.cpp [hunk 2]
```diff
@@ -275,9 +278,30 @@ void QuerySnippet::serializeIntoBuilder(
...
+                TRI_ASSERT(graphNode != nullptr);
+                TRI_ASSERT(graphNode->isDisjoint());
+                // we've found a disjoint smart graph node, now add the `i` th shard for used collections
+                for (auto const& myExp : permuter->second) {
+                  std::string const& cName = myExp.first;
+                  graphNode->addCollectionToShard(cName, *std::next(myExp.second.begin(), i));
+                }
+              }
+            } else {
+              TRI_ASSERT(false);
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:302]: For production please throw an error here after the assert.
It should never happen, but if it happens for some reason there is a chance we end up in UB. later.
Better avoid it.

File: arangod/Aql/QuerySnippet.cpp [hunk 3]
```diff
@@ -275,9 +278,30 @@ void QuerySnippet::serializeIntoBuilder(
         auto permuter = localExpansions.find(current);
         if (permuter != localExpansions.end()) {
           auto collectionAccessingNode = dynamic_cast<CollectionAccessingNode*>(clone);
-          TRI_ASSERT(collectionAccessingNode != nullptr);
-          // Get the `i` th shard
-          collectionAccessingNode->setUsedShard(*std::next(permuter->second.begin(), i));
+          if (collectionAccessingNode != nullptr) {
+            // we guarantee that we only have one collection here
+            TRI_ASSERT(permuter->second.size() == 1);
+            std::string const& cName = collectionAccessingNode->collection()->name();
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:284]: i thank you can can also do a for loop over the permuter->second entry, like you do inside the graph node.

But this should be the same as you do here right now, just saves the lookup of the collection name

File: arangod/Aql/QuerySnippet.cpp [hunk 4]
```diff
@@ -407,14 +441,36 @@ ResultT<std::unordered_map<ExecutionNode*, std::set<ShardID>>> QuerySnippet::pre
...
+        }
+      }
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+      // additional verification checks for disjoint smart graphs
+      if (localGraphNode->isDisjoint()) {
+        size_t size = myExpFinal.begin()->second.size();
+        for (auto const& expDefinition : myExpFinal) {
+          for (auto const& myExp : expDefinition.second) {
+            // 1.) TRI_ASSERT länge der liste = länge der localGraphNode->collections()
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:466]: I gutes this comment is not of much help for men German speaking developers ;)

File: arangod/Aql/QuerySnippet.cpp [hunk 5]
```diff
@@ -407,14 +441,36 @@ ResultT<std::unordered_map<ExecutionNode*, std::set<ShardID>>> QuerySnippet::pre
...
+        if (myExp.size() > 0) {
+          myExpFinal.insert({aqlCollection->name(), std::move(myExp)});
+        }
+      }
+
+#ifdef ARANGODB_ENABLE_MAINTAINER_MODE
+      // additional verification checks for disjoint smart graphs
+      if (localGraphNode->isDisjoint()) {
+        size_t size = myExpFinal.begin()->second.size();
+        for (auto const& expDefinition : myExpFinal) {
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:464]: I think this loop contains wrong assertions.

The assertions should be the following:

// We need one expansion for every collection in the Graph
myExpFinal.size() == localGraphNode->collections().size()
for (exp : myExpFinal) {
  // Every of these expansions needs to have equal amount of shards.
  exp.second.size() == size
}

File: arangod/Aql/QuerySnippet.cpp [hunk 6]
```diff
@@ -474,23 +530,36 @@ ResultT<std::unordered_map<ExecutionNode*, std::set<ShardID>>> QuerySnippet::pre
...
-          // Only in this case we really need to do an expansion
-          // Otherwise we get away with only using the main stream for
-          // this server
-          // NOTE: This might differ between servers.
-          // One server might require an expansion (many shards) while another does not (only one shard).
-          localExpansions.emplace(exp.node, std::move(myExp));
-        }
-      } else {
-        TRI_ASSERT(myExp.size() == 1);
+    }
```
Comment (mchacki) [inline arangod/Aql/QuerySnippet.cpp:533]: I am not sure, but it looks like the scope here is somewhat off.

File: arangod/Aql/QuerySnippet.cpp (hunk 3 above)
Comment (hkernbach) [reply arangod/Aql/QuerySnippet.cpp:284]: Yes, forgot to add a TODO here. Known! But thanks :)

File: arangod/Aql/QuerySnippet.h [hunk 7]
```diff
@@ -81,7 +81,7 @@ class QuerySnippet {
   Id id() const { return _id; }
 
  private:
-  ResultT<std::unordered_map<ExecutionNode*, std::set<ShardID>>> prepareFirstBranch(
+  ResultT<std::unordered_map<ExecutionNode*, std::unordered_map<std::string, std::set<ShardID>>>> prepareFirstBranch(
```
Comment (graetzer) [inline arangod/Aql/QuerySnippet.h:84]: why don't you use a typedef for this stuff?

Comment (mchacki) [reply arangod/Aql/QuerySnippet.h:84]: Yes good point, we can use a typedef here.

**Comment by hkernbach** [comment]
http://jenkins.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/9751/

File: arangod/Aql/QuerySnippet.cpp (hunk 6 above)
Comment (hkernbach) [reply arangod/Aql/QuerySnippet.cpp:568]: Scope is :gut: here :)

File: arangod/Aql/QuerySnippet.cpp (hunk 5 above)
Comment (hkernbach) [reply arangod/Aql/QuerySnippet.cpp:497]: resolved

**Comment by hkernbach** [comment]
http://jenkins01:8080/view/PR/job/arangodb-matrix-pr/9768/

**Pull request merged**